### PR TITLE
Add experimental server-side notification grouping

### PR DESCRIPTION
**Pull request by ClearlyClaire**
This PR adds a `group_key` attribute on the `Notification` model that is filled when creating the notification and is used to allow efficient pagination, fetching no more than one notification per group from the database.

It also adds an experimental API:
- `GET /api/v2_alpha/notifications`
- `GET /api/v2_alpha/notifications/:group_key`
- `POST /api/v2_alpha/notifications/:group_key/dismiss`

This relies on a recursive CTE, which is currently not possible to write with stock ActiveRecord, and therefore relies on a couple monkey-patches:
- a monkey-patch to `ActiveRecord::QueryMethods` to add support for recursive CTEs through a `with_recursive` method, inspired by the one from @lorint (https://github.com/vlado/activerecord-cte/issues/16#issuecomment-1433043310), and submitted in Rails: https://github.com/rails/rails/pull/51601
- a monkey-patch to `Arel::Visitors::ToSql` to address an issue (https://github.com/rails/rails/issues/40181) that caused `arel` to output incorrect SQL for our recursive CTE because of the use of `LIMIT`, now merged upstream: https://github.com/rails/rails/pull/51549

These patches allow to write the CTE in a pretty readable fashion, and allowing the use of already-defined scopes. This can be used to build an API that returns a page of notifications with no more than one notification per group, and still returns a good number of groups, instead of having to call the notification API repeatedly to possibly get more than one group worth of notifications.

## Caveats

- while performance should be really good in most cases, large amounts of notifications within few groups will cause large amounts of rows to be filtered. This may cause performance issues. But it may also work well in practice. Needs more investigation.
- ~~because of the monkey-patches, this introduces a fairly high maintenance burden.
  That being said, one of the two patches has been merged upstream, and the other one is under consideration.~~
  EDIT: both patches have been adopted upstream, so we won't have a long-term maintenance burden
- notification marker ID: we need to return the most recent notification of a page

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
While this PR is still exploratory, this approach is promising. The Arel patch has been accepted into ActiveRecord, so the only monkey-patch left is the recursive CTE one.

On that topic, @lorint, I've used the code you published [here](https://github.com/vlado/activerecord-cte/issues/16#issuecomment-1433043310) to validate the approach, but I have to double-check whether you are ok for this to be included in Mastodon (AGPLv3) or Rails (MIT). Furthermore, are you interested in trying to get this feature merged upstream yourself, or do you prefer I try to do it?

**Comment by ClearlyClaire** [comment]
I think this is ready for review, with the understanding that the API is experimental for now (hence the `/api/v2_alpha` namespace) and will likely change before v4.3.0, so marked it as such.

cc @mjankowski if you can have a quick look at the implementation.

**Comment by mjankowski** [comment]
Some questions before I get into any detailed code review...

- Is the idea here that the `/v1/notifications*` endpoints continue to exist as-is, and these `/v2/...` are opt-in versions for getting grouped notification data? so depending on client need they can either stick with the existing one notification at a time approach or switch to new grouped approach?
- I assume the intention with the namespace is to deploy this in sort of an alpha/preview mode and let clients test against live deployed api, and iterate from there?
- Is there any existing client impl done in tandem to use for guidance here, or is this sort of a first pass at the API level in advance of that? It would be useful to see some (even crude) mocks or wireframes of the expected changing UX to contemplate.

**Comment by ClearlyClaire** [comment]
> * Is the idea here that the `/v1/notifications*` endpoints continue to exist as-is, and these `/v2/...` are opt-in versions for getting grouped notification data? so depending on client need they can either stick with the existing one notification at a time approach or switch to new grouped approach?

Yes. We might discontinue `/api/v1/notifications` eventually, but this is not planned for now.

> * I assume the intention with the namespace is to deploy this in sort of an alpha/preview mode and let clients test against live deployed api, and iterate from there?

Yes.

> * Is there any existing client impl done in tandem to use for guidance here, or is this sort of a first pass at the API level in advance of that? It would be useful to see some (even crude) mocks or wireframes of the expected changing UX to contemplate.

The UI work has not started yet.

**Comment by mjankowski** [comment]
Great, noted on API preview expectations and migration path.

The lack of UI work makes it sort of challenging to review the approach, since there's not a use case or design to support it seems a little speculative, rather than emerging from a need of a feature? If we're doing alpha/preview just to put actual endpoints up for clients to experiment with ... could we literally just hardcode some JSON showing the proposed new response, and get feedback from client devs? Or if there's internal WIP for the web client, leave this in draft until that's further along?

File: app/serializers/rest/notification_group_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class REST::NotificationGroupSerializer < ActiveModel::Serializer
+  attributes :group_key, :notifications_count, :sample_accounts, :type, :group_key
+
+  has_many :sample_accounts, serialized: REST::AccountSerializer
```
Comment (Gargron) [inline app/serializers/rest/notification_group_serializer.rb:6]:
```suggestion
  has_many :sample_accounts, serializer: REST::AccountSerializer
```

File: app/serializers/rest/notification_group_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class REST::NotificationGroupSerializer < ActiveModel::Serializer
+  attributes :group_key, :notifications_count, :sample_accounts, :type, :group_key
```
Comment (Gargron) [inline app/serializers/rest/notification_group_serializer.rb:4]:
```suggestion
  attributes :group_key, :notifications_count, :type
```

**Review by Gargron — DISMISSED** [review]

File: app/models/notification.rb [hunk 3]
```diff
@@ -13,6 +13,7 @@
 #  from_account_id :bigint(8)        not null
 #  type            :string
 #  filtered        :boolean          default(FALSE), not null
+#  group_key       :string
```
Comment (mjankowski) [inline app/models/notification.rb:16]: I wonder about the tradeoffs on the approach here (adding `group_key` string to Notification, the AMS Model, the new NotificationGroup serializer) vs any of:

- Something where the "grouped stats about activity on some object" concept was closer to a first-class object of its own and could be queried/paginated directly instead of going through notifications in that query. Might have similar callback needs to the "Stat" stuff on Status/Account with that approach.
- A materialized view that could have a more direct AR class wrapping up the data.

Do we need to be able to forever query the historical time windows of this data? Like, is there a use case for querying "what did this specific hour three months ago look like?" or is it more real-time/ephemeral and fine if we only have some recent window of data?

File: spec/models/notification_spec.rb [hunk 4]
```diff
@@ -140,6 +140,66 @@
...
+
+    let!(:notifications) do
+      ['group-1', 'group-1', nil, 'group-2', nil, 'group-1', 'group-2', 'group-1']
+        .map { |group_key| Fabricate(:notification, account: account, group_key: group_key) }
+    end
+
+    context 'without since_id or max_id' do
+      it 'returns the most recent notifications, only keeping one notification per group' do
+        expect(described_class.without_suspended.paginate_groups_by_max_id(4).pluck(:id))
+          .to eq [notifications[7].id, notifications[6].id, notifications[4].id, notifications[2].id]
```
Comment (mjankowski) [inline spec/models/notification_spec.rb:154]: This is sort of hard to follow in terms of which array position maps back to which notification in the created set ... might benefit from per-record better named vars?

File: app/models/notification.rb (hunk 3 above)
Comment (ClearlyClaire) [reply app/models/notification.rb:16]: > Something where the "grouped stats about activity on some object" concept was closer to a first-class object of its own and could be queried/paginated directly instead of going through notifications in that query. Might have similar callback needs to the "Stat" stuff on Status/Account with that approach.

Yeah, that is something I have been thinking about as well, but I'm afraid this would involve a lot of rather complex book-keeping open to concurrency issues…

> A materialized view that could have a more direct AR class wrapping up the data.

I am not too sure how that would look like, and materialized views, with `Scenic` in particular, have proved to be a massive pain in handling migration paths.

> Do we need to be able to forever query the historical time windows of this data? Like, is there a use case for querying "what did this specific hour three months ago look like?" or is it more real-time/ephemeral and fine if we only have some recent window of data?

Do you mean outright losing old notifications, or losing the way they are grouped?

File: spec/models/notification_spec.rb (hunk 4 above)
Comment (ClearlyClaire) [reply spec/models/notification_spec.rb:154]: yeah, that would make sense

File: app/models/notification.rb (hunk 3 above)
Comment (mjankowski) [reply app/models/notification.rb:16]: > Do you mean outright losing old notifications, or losing the way they are grouped?

More the latter. Like from the api perspective, can you keep paginating/querying infinitely far back (or I guess, up to the introduction of this feature) and expect to always have results? This is sort of in the context of the DB view approach, and whether you could have it built for a moving window time period (last X days?) which might keep the view generation more modest, easier to keep current, etc.

File: app/services/notify_service.rb [hunk 5]
```diff
@@ -177,6 +180,7 @@ def call(recipient, type, activity)
     return if dismiss?
 
     @notification.filtered = filter?
+    @notification.group_key = notification_group_key
```
Comment (mjankowski) [inline app/services/notify_service.rb:183]: Are there scenarios where we want to create `Notification` records but NOT generate the group key? If yes, this is fine -- if no, could/should this be a callback on the model?

File: app/services/notify_service.rb [hunk 6]
```diff
@@ -196,6 +200,24 @@ def call(recipient, type, activity)
 
   private
 
+  def notification_group_key
```
Comment (mjankowski) [inline app/services/notify_service.rb:203]: A good chunk of this method is dealing with attributes on `Notification`, and the parts that are not are managing the hour bucket concept and how it interacts with redis. I wonder if some/all of:

- Move this method onto `Notification` directly? (may need to pass in the recipient, as currently done)
- Add a class which manages the redis querying/checking/setting/key aspects?
- Keep this in the notify service, but turn this one larger method into 2-3 smaller methods which separate the different things being done here

File: spec/requests/api/v2_alpha/notifications_spec.rb [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    end
+
+    it_behaves_like 'forbidden for wrong scope', 'write write:notifications'
+
+    context 'with no options' do
+      it 'returns expected notification types', :aggregate_failures do
+        subject
+
+        expect(response).to have_http_status(200)
+        expect(body_json_types).to include 'reblog'
```
Comment (mjankowski) [inline spec/requests/api/v2_alpha/notifications_spec.rb:37]: Super minor style thing here -- I think you can pass multiple values to `include` (ie, `include('reblog', 'mention', ...)` and it will pass as long as they are all in the checked value ... or if not that, you could chain together multiple checks with `.and include(...)` to keep them all in one `expect`.

File: app/models/notification.rb (hunk 3 above)
Comment (ClearlyClaire) [reply app/models/notification.rb:16]: I'm honestly not sure. I think if grouping is important to keep things readable, it remains important for the past as well. Possibly we could allow coarser grouping for older entries, but I don't think it would make things easier.

File: app/services/notify_service.rb (hunk 5 above)
Comment (ClearlyClaire) [reply app/services/notify_service.rb:183]: I can't find of any such scenario. I did it in the service rather than the model for a few reasons, namely:
- it is the only place where we need to do it (it's not a strong reason to not do it in the model though)
- it was initially more complex (not that relevant anymore)
- it initially relied on the `id` of the created notification (that is not relevant anymore)
- it uses Redis in a way the model doesn't afaik

Those reasons have been greatly weakened since, so I'll consider moving it to the model.

**Comment by renchap** [comment]
There is a TODO left:
> # TODO: most recent and/or oldest notification ID

I started implementing it on the frontend, let's discuss how pagination / gap filling should work so we can add this and ensure everything is there to handle those use-cases

**Comment by ClearlyClaire** [comment]
> There is a TODO left:
> 
> > TODO: most recent and/or oldest notification ID
> 
> I started implementing it on the frontend, let's discuss how pagination / gap filling should work so we can add this and ensure everything is there to handle those use-cases

Basically, a grouped notification page spans a range of notification IDs, and paginating efficiently requires continuing from the cutoff. But looking at this API again, that range is only exposed through pre-built URLs in the `Link` headers, which is exceptionally awkward to deal with if you need more than exploring pages as exposed by this API (e.g. you handle notifications from another source).

The ID span of notification groups *within the page* would be one way to expose that information, although that would be confusing, since they would not be a property of the group itself, but of the group and the page…

A related issue is that the CTE will not return such a span for each group, but only one end of this span (the first matching notification for each group, not the notifications which have been ignored because another one form the same group was already considered).

**Comment by ClearlyClaire** [comment]
I added `page_min_id` and `page_max_id` to `NotificationGroup` when a page is requested, in order to cover the above use-case (the effective range is the `page_min_id` of the last group to the `page_max_id` of the first group in the page) as well as better track which is the most recent notification ID read by a user (`max_page_id` of a group).

I have not added tests yet, and I have not tested performances of that extra database query.

I suspect having `page_min_id` for each group is actually not useful and some database work could be avoided by wrapping the array in an object with additional attributes `min_id` and `max_id`.

**Review by renchap — DISMISSED** [review]
I am fine with merging this.

The database side should not change now, and the API is marked as experimental, so I do not expect problems deploying it.

We can change the API in further PRs until we mark it as stable if needed, and having it merged will make working on the frontend implementation much easier and less prone to rebase issues.

File: spec/models/notification_spec.rb (hunk 4 above)
Comment (ClearlyClaire) [reply spec/models/notification_spec.rb:165]: I tried doing that, using multiple `let!` to enforce the correct order, with objects named `group1_notif1`, `group1_notif2`, `ungrouped_notif1` and so on, which does improve readability but contradicts multiple of the rubocop rules. I'm not sure which is preferable, so postponing this.

**Comment by ClearlyClaire** [comment]
@mjankowski I marked your conversations as resolved in order to move forward, but there's a couple refactors that may still be worth exploring:
- moving the `notification_group_key` logic from `NotifyService` to the `Notification` class
- refactor pagination tests to use named variables (but this requires fighting rubocop)

**Review by Gargron — APPROVED** [review]
